### PR TITLE
mutation_fragment_stream_validator: various API improvements

### PR DESCRIPTION
**Pull request by denesb**
The low-level `mutation_fragment_stream_validator` gets `reset()` methods that until now only the high-level `mutation_fragment_stream_validating_filter` had.
Active tombstone validation is pushed down to the low level validator.
The low level validator, which was a pain to use until now due to being very fussy on which subset of its API one used, is made much more robust, not requiring the user to stick to a subset of its API anymore.

### CODE REVIEW

File: readers/mutation_reader.cc [hunk 1]
```diff
@@ -39,111 +41,149 @@ mutation_fragment_stream_validator::mutation_fragment_stream_validator(const ::s
...
-}
+    auto valid = true;
 
-bool mutation_fragment_stream_validator::operator()(mutation_fragment_v2::kind kind) {
-    bool valid = true;
+    // Check fragment kind order
     switch (_prev_kind) {
         case mutation_fragment_v2::kind::partition_start:
-            valid = kind != mutation_fragment_v2::kind::partition_start;
+            valid &= kind != mutation_fragment_v2::kind::partition_start;
```
Comment (bhalevy) [inline readers/mutation_reader.cc:86]: why the change to `&=`?

File: readers/mutation_reader.cc [hunk 2]
```diff
@@ -39,111 +41,149 @@ mutation_fragment_stream_validator::mutation_fragment_stream_validator(const ::s
...
+        return false;
+    }
+
+    if (pos && _prev_kind != mutation_fragment_v2::kind::partition_end) {
+        auto cmp = position_in_partition::tri_compare(_schema);
+        auto res = cmp(_prev_pos, *pos);
+        if (_prev_kind == mutation_fragment_v2::kind::range_tombstone_change) {
+            valid &= res <= 0;
+        } else {
+            valid &= res < 0;
```
Comment (bhalevy) [inline readers/mutation_reader.cc:108]: here too it doesn't seem like we need `valid &=`

File: readers/mutation_reader.cc [hunk 3]
```diff
@@ -39,111 +41,149 @@ mutation_fragment_stream_validator::mutation_fragment_stream_validator(const ::s
...
+        switch (kind) {
+            case mutation_fragment_v2::kind::partition_start:
+                _prev_pos = position_in_partition(position_in_partition::partition_start_tag_t{});
+                break;
+            case mutation_fragment_v2::kind::static_row:
+                _prev_pos = position_in_partition(position_in_partition::static_row_tag_t{});
+                break;
+            case mutation_fragment_v2::kind::clustering_row:
+            case mutation_fragment_v2::kind::range_tombstone_change: // fall-through
+                if (!_prev_pos.has_clustering_key()) { // don't move pos if it is already a clustering one
```
Comment (bhalevy) [inline readers/mutation_reader.cc:128]: so how the current `pos` ever gets into prev_pos?

File: readers/mutation_reader.cc [hunk 4]
```diff
@@ -39,111 +41,149 @@ mutation_fragment_stream_validator::mutation_fragment_stream_validator(const ::s
...
+                    _prev_pos = position_in_partition(position_in_partition::before_clustering_row_tag_t{}, clustering_key::make_empty());
+                }
+                break;
+            case mutation_fragment_v2::kind::partition_end:
+                _prev_pos = position_in_partition(position_in_partition::end_of_partition_tag_t{});
+                break;
+        }
     }
-    return valid;
+    _current_tombstone = new_current_tombstone.value_or(_current_tombstone);
```
Comment (bhalevy) [inline readers/mutation_reader.cc:137]: why not prevent writing to memory if new_current_tombstone is disengaged?
this is on the hot path and value_or doesn't save the conditional anyhow...

File: readers/mutation_reader.cc [hunk 5]
```diff
@@ -236,18 +275,35 @@ bool mutation_fragment_stream_validating_filter::operator()(mutation_fragment_v2
 }
 
 bool mutation_fragment_stream_validating_filter::operator()(mutation_fragment::kind kind, position_in_partition_view pos) {
-    return (*this)(to_mutation_fragment_kind_v2(kind), pos);
+    return (*this)(to_mutation_fragment_kind_v2(kind), pos, {});
 }
```
Comment (bhalevy) [inline readers/mutation_reader.cc:279]: nit: std::nullopt is clearer IMO

File: readers/mutation_reader.cc [hunk 6]
```diff
@@ -236,18 +275,35 @@ bool mutation_fragment_stream_validating_filter::operator()(mutation_fragment_v2
...
 
 bool mutation_fragment_stream_validating_filter::operator()(const mutation_fragment_v2& mv) {
-    auto valid = (*this)(mv.mutation_fragment_kind(), mv.position());
-    if (valid && mv.is_range_tombstone_change()) {
-        _current_tombstone = mv.as_range_tombstone_change().tombstone();
-    }
-    return valid;
+    return (*this)(mv.mutation_fragment_kind(), mv.position(),
+            mv.is_range_tombstone_change() ? std::optional(mv.as_range_tombstone_change().tombstone()) : std::nullopt);
 }
```
Comment (bhalevy) [inline readers/mutation_reader.cc:284]: How about defining the this as a mutation_fragment_v2 method?

**Review by bhalevy — CHANGES_REQUESTED** [review]

File: readers/mutation_reader.cc (hunk 1 above)
Comment (denesb) [reply readers/mutation_reader.cc:86]: We no longer return as soon as find something invalid so we have to be careful not to overwrite a previous false with a true. Maybe not all these are strictly needed, I just wanted to be careful.

File: readers/mutation_reader.cc (hunk 2 above)
Comment (denesb) [reply readers/mutation_reader.cc:108]: See above.

File: readers/mutation_reader.cc (hunk 3 above)
Comment (denesb) [reply readers/mutation_reader.cc:128]: See the true branch above: if `pos` is engaged it always gets into `_prev_pos`.

File: readers/mutation_reader.cc (hunk 6 above)
Comment (denesb) [reply readers/mutation_reader.cc:284]: I don't think it makes sense to do so to save an if on a single user. `mutation_fragment_v2` should have getters only for things that are common among all mutation fragment kinds, like `position()`.

File: readers/mutation_reader.cc (hunk 4 above)
Comment (denesb) [reply readers/mutation_reader.cc:137]: I was optimizing for number of source lines. I agree an if here makes more sense.

File: readers/mutation_reader.cc (hunk 1 above)
Comment (denesb) [reply readers/mutation_reader.cc:86]: Actually, on second look, we don't need the `&=`. We do have a return check after each validation step.

**Comment by denesb** [comment]
New in v2:
* Rebased.
* Remove unnecessary `valid &=` assignments, simple `valid = ` is enough.
* Don't write to `_current_tombstone` when there is no update.

File: readers/mutation_reader.cc (hunk 1 above)
Comment (bhalevy) [reply readers/mutation_reader.cc]: > Actually, on second look, we don't need the `&=`. We do have a return check after each validation step.

exactly.

File: readers/mutation_reader.cc (hunk 2 above)
Comment (bhalevy) [reply readers/mutation_reader.cc:108]: we return immediately below `if (!valid)`, so there's no need for `&=` here too.

File: readers/mutation_reader.cc (hunk 3 above)
Comment (bhalevy) [reply readers/mutation_reader.cc:128]: I presume this would handle dummy rows.

Comment (denesb) [reply readers/mutation_reader.cc:128]: Actually it is better to check for partition-region instead here. Not all clustering positions have a key.

File: readers/mutation_reader.cc [hunk 7]
```diff
@@ -112,38 +95,97 @@ bool mutation_fragment_stream_validator::operator()(mutation_fragment_v2::kind k
...
+    } else {
+        switch (kind) {
+            case mutation_fragment_v2::kind::partition_start:
+                _prev_pos = position_in_partition(position_in_partition::partition_start_tag_t{});
+                break;
+            case mutation_fragment_v2::kind::static_row:
+                _prev_pos = position_in_partition(position_in_partition::static_row_tag_t{});
+                break;
+            case mutation_fragment_v2::kind::clustering_row:
+            case mutation_fragment_v2::kind::range_tombstone_change: // fall-through
```
Comment (bhalevy) [inline readers/mutation_reader.cc:127]: nit: use `[[fallthrough]];` on the case above (that falls) through, if at all.
https://en.cppreference.com/w/cpp/language/attributes/fallthrough
That said, I'm not sure this is really needed for a classic sequence of case labels like here.

File: readers/mutation_reader.cc (hunk 6 above)
Comment (bhalevy) [reply readers/mutation_reader.cc:286]: FWIW, there's another user in `mutation_fragment_stream_validator::reset`.
I would have suggested to define the getter that returns an optional range_tombstone their if this was the same class.

File: readers/mutation_reader.cc [hunk 8]
```diff
@@ -236,18 +277,35 @@ bool mutation_fragment_stream_validating_filter::operator()(mutation_fragment_v2
...
-    if (valid && mv.is_range_tombstone_change()) {
-        _current_tombstone = mv.as_range_tombstone_change().tombstone();
-    }
-    return valid;
+    return (*this)(mv.mutation_fragment_kind(), mv.position(),
+            mv.is_range_tombstone_change() ? std::optional(mv.as_range_tombstone_change().tombstone()) : std::nullopt);
 }
 bool mutation_fragment_stream_validating_filter::operator()(const mutation_fragment& mv) {
-    return (*this)(to_mutation_fragment_kind_v2(mv.mutation_fragment_kind()), mv.position());
+    return (*this)(to_mutation_fragment_kind_v2(mv.mutation_fragment_kind()), mv.position(), {});
```
Comment (bhalevy) [inline readers/mutation_reader.cc:288]: nit: `std::nullopt` rather than `{}`

File: readers/mutation_reader.cc (hunk 3 above)
Comment (bhalevy) [reply readers/mutation_reader.cc:128]: agreed

File: readers/mutation_reader.cc (hunk 8 above)
Comment (denesb) [reply readers/mutation_reader.cc:289]: I think `{}` already conveys the purpose: a parameter whose value we don't care about. We use this in many places in the code.

**Comment by denesb** [comment]
New in v3:
* Use `[[fallthrough]]` instead of comment.
* Check for partition-region instead of for clustering key (not all clustering elements will have a key).

File: readers/mutation_reader.cc (hunk 2 above)
Comment (bhalevy) [reply readers/mutation_reader.cc:108]: @denesb why is `&=` needed here?
`valid` always starts as `true` before entering this block and we always return immediately if it's `false` after the block.

**Review by bhalevy — CHANGES_REQUESTED** [review]
The only material comment left is about the remaining `valid &=` logic.

**Comment by denesb** [comment]
> The only material comment left is about the remaining `valid &=` logic.

Ah, I see now I forgot about one spot. Will fix.

**Comment by denesb** [comment]
New in v4:
* Removed forgotten instances of `valid &=`.

**Review by bhalevy — APPROVED** [review]

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2390/

[Sanity Tests / resharding_test.TestReshardingSingleNodeGating.test_resharding_by_murmur3_gating[1-TimeWindowCompactionStrategy-15]](https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2390/testReport/junit/resharding_test/TestReshardingSingleNodeGating/Sanity_Tests___test_resharding_by_murmur3_gating_1_TimeWindowCompactionStrategy_15_/) failed, the test timed out. 
Unrelated to this PR, restarting the CI

**Comment by denesb** [comment]
Ping.

**Comment by denesb** [comment]
@scylladb/scylla-maint ping.

File: readers/mutation_reader.cc [hunk 9]
```diff
@@ -226,6 +226,9 @@ bool mutation_fragment_stream_validating_filter::operator()(mutation_fragment_v2
         } else if (_validation_level >= mutation_fragment_stream_validation_level::partition_key) {
             on_validation_error(mrlog, format("[validator {} for {}] Unexpected mutation fragment: partition key {}: previous {}, current {}",
                     static_cast<void*>(this), _name, _validator.previous_partition_key(), _validator.previous_mutation_fragment_kind(), kind));
+        } else if (kind == mutation_fragment_v2::kind::partition_end && _validator.current_tombstone()) {
```
Comment (tgrabiec) [inline readers/mutation_reader.cc:268]: Unrelated to this patch, but do we try to recreate the reason here, rather than have the _validator return the reason for validation error? Seems like this is should be the responsibility of the low level validator.

Comment (denesb) [reply readers/mutation_reader.cc:268]: Good point. I find myself composing the same messages over and over again. It would be better if the low-level validator did this.

Comment (denesb) [reply readers/mutation_reader.cc:268]: I will do this in a follow-up.

**Comment by denesb** [comment]
@sclladb/scylla-maint ping. This PR is blocking progress on another PR and soon yet another one.

**Comment by bhalevy** [comment]
> @sclladb/scylla-maint ping.

@scylladb/scylla-maint ping (typo above)
